### ISSUE TITLE
Support resuming of image/diff pulls

### ISSUE THREAD

**Issue by jakogut** [balena-os/balena-engine#249]
If the engine is pulling an image or diff, and the connection is interrupted, the download will restart from the beginning instead of resuming from the last chunk of data that was downloaded successfully.

Steps to reproduce:
1. Setup any device with OS v2.80.0 or older
2. Deploy an application release large enough to not download immediately, such as 1 GB or more
3. As the device is downloading the image, disable its connection to the internet. If it's setup to use a Linux machine as a router, you can do this with `sysctl -w net.ipv4.ip_forward=0`
4. Wait a few seconds, and re-enable the connection. The logs will show that the connection was reset, as well as: `Delta failed with Error: failed to register layer: Error processing tar file(exit status 1): unexpected EOF`

Layer diffs reduce the download bandwidth of deploying releases, but large diffs on unreliable networks can still be onerous and expensive to deploy, as any interruptions force it to start from the beginning. Instead, the engine should download the image tarball in chunks, and resume the download from the last successful chunk when possible.

**Comment by jakogut** [comment]
Related: https://github.com/distribution/distribution/issues/762
https://github.com/moby/moby/pull/28348

**Comment by joehounsham** [comment]
@jakogut Do you have some estimations as to when we can expect this to be deployed?

**Comment by jakogut** [comment]
@lmbarros Is working on this, my understanding is that he's working on some tests to ensure there are no regressions.

**Comment by lmbarros** [comment]
@joehounsham Resuming image downloads after network failures were implemented by this PR: https://github.com/balena-os/balena-engine/pull/253. We might need to adjust things like the number of times we retry and the amount of time we wait between retries, but otherwise I hope this to be a good improvement.

It's available in [balenaEngine 19.03.23](https://github.com/balena-os/balena-engine/blob/master/CHANGELOG.md#v190323) and [balenaOS 2.82.7](https://github.com/balena-os/meta-balena/blob/master/CHANGELOG.md#v2827). I am not sure this OS version is available in production for any device type, but I know we released it on staging for Xavier NX SD-CARD. I am aware of one customer trying it currently on staging and AFAIK, so far so good.
